### ISSUE TITLE
Order sometimes incorrect

### ISSUE THREAD

**Issue by pndewit** [michaelmalonenz/aurelia-dragula#45]
As already mentioned in [my comment](https://github.com/michaelmalonenz/aurelia-dragula/pull/44#issuecomment-383009215) I was running into issues with plain Dragula which caused the drop and dragend events to be emitted before all DOM changes were completed, making it impossible to keep track of the order they were in. I replaced the module with aurelia-dragula and was hoping it would fix it, but apparently it didn't completely. I am still getting an incorrect order sometimes, but even less than with plain Dragula. The most frustrating part is that I have no idea how to reproduce it and thus what the root cause of this issue is.
Hope you might be able to help me in some way.

    drop(item, target, source, sibling, itemVM, siblingVM) {
        let siblingResource = siblingVM ? siblingVM.resource : null;
        moveBefore(this.resourceOrder, resource => resource === itemVM.resource, resource => resource === siblingResource);
    }

    <dragula-and-drop drop-fn.call="drop(item, target, source, sibling, itemVM, siblingVM)"></dragula-and-drop>
    <ul class="drop-target drag-source}">
        <li repeat.for="resource of resourceOrder" resource.bind="resource">${resource}</li>
    </ul>

Some sidenotes that may be interesting:
- revertOnSpill is enabled (https://github.com/michaelmalonenz/aurelia-dragula/pull/44)
- direction is vertical (default)

**Comment by pndewit** [comment]
False alarm, I was making a mistake myself...

**Comment by pndewit** [comment]
I did encounter a real incorrect order issue now and I think it has something to do with the `<!-- anchor -->` comment nodes that [Aurelia is creating](https://github.com/aurelia/templating/issues/57). The `_initialSibling` is sometimes [referring to such a comment node](https://github.com/michaelmalonenz/aurelia-dragula/blob/master/src/dragula.js#L208) making it a [different sibling then `_currentSibling`](https://github.com/michaelmalonenz/aurelia-dragula/blob/master/src/dragula.js#L324).

**Comment by pndewit** [comment]
Updating [L208](https://github.com/michaelmalonenz/aurelia-dragula/blob/master/src/dragula.js#L208) to match with [L209](https://github.com/michaelmalonenz/aurelia-dragula/blob/master/src/dragula.js#L209) from:
`this._initialSibling = context.item.nextSibling;`

to:
`this._initialSibling = Util.nextEl(context.item);`

Fixes the issue where the `_initialSibling` and the `_currentSibling` are incorrectly being marked as not the same, so it is not throwing incorrect `drop` events any more. It is still not fixing my issue where the order is sometimes incorrect. This might be in the same area though!
I'll so some more research and maybe create a PR if I find a solution.

**Comment by michaelmalonenz** [comment]
One of the tipping points for me creating this library as opposed to using the vanilla `dragula` library was getting the order handling working within the Aurelia environment - which means I've looked at this scenario pretty closely, so I'd appreciate a reproducer if you believe there's a problem in here.

Thanks!

**Comment by pndewit** [comment]
Sure! Using your example project: https://github.com/michaelmalonenz/aurelia-dragula-example
1) Add a breakpoint, console log, or anything on line 32 in `welcome.js` (inside the `drop` function)
2) Start it up and submit the name: "test"
3) Note that the repeater DOM ends with a `<!--anchor-->` node (after the last `<compose>` element
4) Move the last (right) letter, but do not actually change its position in the list. Just drag it, move it a px or 2 to initiate the drag and release it again.

The drop event should now have fired! Like I said this is because the `_initialSibling` is set to the `<!--anchor-->` from step 3 which is then compared to the `_currentSibling` which is set to `null`. They are obviously not equal, triggering the `drop` event.

**side note**, did you know your example project hangs your browser tab if you submit with a blank name? 😆

**Comment by pndewit** [comment]
### Problem 2
This one is more serious and also related to the `<!--anchor-->` node. Using your example project: https://github.com/michaelmalonenz/aurelia-dragula-example

1) Start it up and submit a name with all different letter, e.g.: "abcd"
2) Note that the repeater DOM ends with a `<!--anchor-->` node (after the last `<compose>` element)
3) Move the last (right) letter, but do not actually change its position in the list. Just drag it, move it a px or 2 to initiate the drag and release it again.
4) Note that the repeater DOM now ends with a `<compose>` element and the `<!--anchor-->` node has now moved up one sibling. If the `<!--anchor-->` node is still in the last position (happens sometimes, no clue why), try moving it again but now hover it to another position in the list and move it back to its original position and drop it. For me this triggered the `<!--anchor-->` node to be moved as well.
5) You might have guessed it already by now, but when you now move the 3rd letter to the last (4th) position, it moves back to the original position (position 3)! This sometimes also triggers the `<!--anchor-->` node to be moved one up again.

**Comment by pndewit** [comment]
I've done some research for problem 2 and it seems like it is caused because `aurelia-templating` and `aurelia-binding` calculate that there are no differences in the array and thus do nothing to the DOM. But `Dragula` actually did change something (move the `<compose>` element after the `<!--anchor-->` element). With a normal change (drop) `aurelia-templating` and `aurelia-binding` would then move the `<!--anchor-->` element to the end again.

**Comment by michaelmalonenz** [comment]
Righto - I've definitely seen your issue as described, but I disagree that using the `Util.nextEl` function is the answer.  I will have a bit of a think about it and come up with a solution if I can.

**Comment by pndewit** [comment]
Fair enough, can I ask why this is not the solution for the first problem? Maybe I'll be able to help you :)

Have you also seen the second issue in action?

**Comment by michaelmalonenz** [comment]
I was referring to the second issue - I haven't seen the first.  I think they are both symptoms of the same issue.

Crap.  I know the problem, but not (necessarily how to fix it).
When I fixed the flicker on drop, I stopped inserting before the `<!-- anchor -->` element on revert - the _initialSibling variable came into being in order to fix this issue that exists in plain dragula (because it doesn't cater to Aurelia's comment elements)  

This is the condition:
https://github.com/michaelmalonenz/aurelia-dragula/blob/master/src/dragula.js#L288

If you change it to:
`    if (initial === false && !this._copy && reverts) {`

then you'll see your problems disappear
